src/pipeline/RADEReceiveStep.cpp

```diff
@@ -0,0 +1,141 @@
+//=========================================================================
+// Name:            RADEReceiveStep.cpp
+// Purpose:         Describes a demodulation step in the audio pipeline.
+//
+// Authors:         Mooneer Salem
+// License:
+//
+//  All rights reserved.
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.1,
+//  as published by the Free Software Foundation.  This program is
+//  distributed in the hope that it will be useful, but WITHOUT ANY
+//  WARRANTY; without even the implied warranty of MERCHANTABILITY or
+//  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+//  License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, see <http://www.gnu.org/licenses/>.
+//
+//=========================================================================
+
+#include <cassert>
+#include "RADEReceiveStep.h"
+#include "../defines.h"
+#include "lpcnet.h" // from Opus source tree
+
+RADEReceiveStep::RADEReceiveStep(struct rade* dv, FARGANState* fargan)
+    : dv_(dv)
+    , fargan_(fargan)
+    , inputSampleFifo_(nullptr)
+    , outputSampleFifo_(nullptr)
+{
+    // Set FIFO to be 2x the number of samples per run so we don't lose anything.
+    inputSampleFifo_ = codec2_fifo_create(rade_nin_max(dv_) * 2);
+    assert(inputSampleFifo_ != nullptr);
+
+    // Enough for one second of audio. Probably way overkill.
+    outputSampleFifo_ = codec2_fifo_create(16000);
+    assert(outputSampleFifo_ != nullptr);
+}
+
+RADEReceiveStep::~RADEReceiveStep()
+{
+    if (inputSampleFifo_ != nullptr)
+    {
+        codec2_fifo_free(inputSampleFifo_);
+    }
+
+    if (outputSampleFifo_ != nullptr)
+    {
+        codec2_fifo_free(outputSampleFifo_);
+    }
+}
+
+int RADEReceiveStep::getInputSampleRate() const
+{
+    return 8000; 
+}
+
+int RADEReceiveStep::getOutputSampleRate() const
+{
+    return 16000;
+}
+
+std::shared_ptr<short> RADEReceiveStep::execute(std::shared_ptr<short> inputSamples, int numInputSamples, int* numOutputSamples)
```

Comment:
@drowe67, can you take a look at this function and let me know if it seems reasonable? I'm getting the following segfault on Linux after I push Start in the freedv-gui built by this branch:

```
encoder: 937200 weights
decoder: 907764 weights
encoder: 937200 weights
decoder: 907764 weights
Rs: 33.33 Rs': 50.00 Ts': 0.020 Nsmf: 120 Ns:   4 Nc:  30 M: 160 Ncp: 32
Input BPF bandwidth: 1740.000244 centre: 1475.000000
n_features_out: 432 nin_max: 1120 nin: 960
Setting callsign to K6AQ
[New Thread 0x7fff71a006c0 (LWP 4734)]
[Thread 0x7fff71a006c0 (LWP 4734) exited]
[New Thread 0x7fff71a006c0 (LWP 4735)]
[New Thread 0x7fff710006c0 (LWP 4736)]
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to record device alsa_input.pci-0000_02_02.0.analog-stereo
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to playback device alsa_output.platform-snd_aloop.0.analog-stereo
[New Thread 0x7fff63e006c0 (LWP 4737)]
[New Thread 0x7fff634006c0 (LWP 4738)]
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to record device alsa_input.platform-snd_aloop.0.analog-stereo
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to playback device alsa_output.pci-0000_02_02.0.analog-stereo
[New Thread 0x7fff62a006c0 (LWP 4739)]
[Thread 0x7fffc7e006c0 (LWP 4719) exited]
txRxThread: timeout while waiting for CV, tx = 1
txRxThread: timeout while waiting for CV, tx = 0
txRxThread: timeout while waiting for CV, tx = 1
txRxThread: timeout while waiting for CV, tx = 0
txRxThread: timeout while waiting for CV, tx = 1

Thread 31 "FreeDV rxThread" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff634006c0 (LWP 4738)]
0x00007ffff44f78d5 in ?? () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
(gdb) bt
#0  0x00007ffff44f78d5 in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#1  0x00007ffff449c936 in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#2  0x00007ffff449d356 in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#3  0x00007ffff438604d in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#4  0x00007ffff7e5f457 in rade_rx ()
    at /home/mooneer/freedv-gui/build_linux/rade_build/src/librade.so.0.1
#5  0x000055555565feb7 in RADEReceiveStep::execute
    (this=0x7fff38010a60, inputSamples=std::shared_ptr<short> (use count 6, weak count 0) = {...}, numInputSamples=141, numOutputSamples=0x7fff633fe6ac)
    at /home/mooneer/freedv-gui/src/pipeline/RADEReceiveStep.cpp:95
#6  0x0000555555650629 in operator()
    (__closure=0x7fff380390f8, inSamples=std::shared_ptr<short> (use count 6, weak count 0) = {...}, numSamples=160)
    at /home/mooneer/freedv-gui/src/pipeline/ParallelStep.cpp:215
#7  0x0000555555651cc0 in std::__invoke_impl<std::pair<std::shared_ptr<short int>, int>, ParallelStep::enqueueTask_(ThreadInfo*, IPipelineStep*, std::shared_ptr<short int>, int)::<lambda(std::shared_ptr<short int>, int)>&, std::shared_ptr<short int>, int>(std::__invoke_other, struct {...} &) (__f=...)
    at /usr/include/c++/13/bits/invoke.h:61
#8  0x0000555555651a7c in std::__invoke_r<std::pair<std::shared_ptr<short int>, int>, ParallelStep::enqueueTask_(ThreadInfo*, IPipelineStep*, std::shared_ptr<short int>, int)::<lambda(std::shared_ptr<short int>, int)>&, std::shared_ptr<short int>, int>(struct {...} &) (__fn=...)
--Type <RET> for more, q to quit, c to continue without paging--q
```

I'm executing by running the following:

```
mooneer@ubuntu-vm:~/freedv-gui/build_linux$ . ~/radae/radae-venv/bin/activate
(radae-venv) mooneer@ubuntu-vm:~/freedv-gui/build_linux$ PYTHONPATH=rade_src/embed:rade_src/ src/freedv
```

Reply:
Actually, looks like it was a threading issue. See https://github.com/drowe67/radae/pull/27/commits/300f2e302298577efac78748c941a3d5433a931b.

Unfortunately RADE seems to be behaving pretty slowly in the Ubuntu VM I'm using for some reason. I need to investigate this more.